util/src/org/commcare/util/CommCareConfigEngine.java

```diff
@@ -379,4 +382,123 @@ private void print(Suite s, Entry e, int level) {
             }
         }
     }
+
+
+    final static private class QuickStateListener implements TableStateListener{
+        int lastComplete = 0;
+        int lastTotal = 0;
+
+        @Override
+        public void resourceStateUpdated(ResourceTable table) {
+
+        }
+
+        @Override
+        public void incrementProgress(int complete, int total) {
+            int diff = complete - lastComplete;
+            lastComplete = complete;
+            for(int i = 0 ; i < diff ; ++i) {
+                System.out.print(".");
+            }
+        }
+    };
+
+    public void attemptAppUpdate(boolean forceNew) {
+        ResourceTable global = table;
+
+        // Ok, should figure out what the state of this bad boy is.
+        Resource profileRef = global.getResourceWithId(CommCarePlatform.APP_PROFILE_RESOURCE_ID);
+
+        Profile profileObj = this.getPlatform().getCurrentProfile();
+
+        boolean appInstalled = (profileRef != null &&
+                profileRef.getStatus() == Resource.RESOURCE_STATUS_INSTALLED);
+
+        TableStateListener globalListener = new TableStateListener() {
+            int lastComplete = 0;
+
+            @Override
+            public void resourceStateUpdated(ResourceTable table) {
+
+            }
+
+            @Override
+            public void incrementProgress(int complete, int total) {
+
+            }
+        };
+
+        global.setStateListener(new QuickStateListener());
+
+        updateTable.setStateListener(new QuickStateListener());
+
+        // When profileRef points is http, add appropriate dev flags
+        String authRef = profileObj.getAuthReference();
+
+        try {
+            URL authUrl = new URL(authRef);
+
+            // profileRef couldn't be parsed as a URL, so don't worry
+            // about adding dev flags to the url's query
+
+            // If we want to be using/updating to the latest build of the
+            // app (instead of latest release), add it to the query tags of
+            // the profile reference
+            if (forceNew &&
+                    (authUrl != null) &&
+                    ("https".equals(authUrl.getProtocol()) ||
+                            "http".equals(authUrl.getProtocol()))) {
+                if (authUrl.getQuery() != null) {
+                    // If the profileRef url already have query strings
+                    // just add a new one to the end
+                    authRef = authRef + "&target=build";
+                } else {
+                    // otherwise, start off the query string with a ?
+                    authRef = authRef + "?target=build";
+                }
```

Comment:
they use different URL objects unfortunately, but yes, we should definitely find a wy to centralize this logic